server/src/main/java/org/elasticsearch/index/mapper/DocumentParserContext.java

```diff
@@ -103,24 +104,23 @@ private DocumentParserContext(
         SourceToParse sourceToParse,
         Set<String> ignoreFields,
         Map<String, List<Mapper>> dynamicMappers,
-        Set<String> newFieldsSeen,
         Map<String, ObjectMapper> dynamicObjectMappers,
-        List<RuntimeField> dynamicRuntimeFields,
+        Map<String, List<RuntimeField>> dynamicRuntimeFields,
         String id,
         Field version,
         SeqNoFieldMapper.SequenceIDFields seqID,
         DocumentDimensions dimensions,
         ObjectMapper parent,
         ObjectMapper.Dynamic dynamic,
         Set<String> fieldsAppliedFromTemplates,
-        Set<String> copyToFields
+        Set<String> copyToFields,
+        AtomicInteger dynamicMapperSize
```

Comment:
It doesn't need to be atomic but it needs to be a mutable integer value that's shared across multiple instances of `DocumentParserContext` so that the child contexts created via `DocumentParserContext#createChildContext` increment the same value.

Reply:
If you think it's more readable I could create a simple `MutableInteger` class. It could either be in some common or util package or just private to `DocumentParserContext`.

Reply:
yes, I think it's confusing to use atomic data structures for non concurrent scenarios. i would try to hide this behind the context object and make it a private static inner class if possible.

Reply:
I've replaces this with a `private static final class DynamicMapperSize`